hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/replication/ContinuousBackupStagedHFileCleaner.java

```diff
@@ -0,0 +1,148 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.collect.Iterables;
+
+/**
+ * A custom HFile cleaner delegate for continuous backup scenarios in HBase. This cleaner prevents
+ * the deletion of HFiles that are staged for bulk loading as part of the continuous backup process.
+ * It interacts with the HBase `StagedBulkloadFileRegistry` to determine which files should be
+ * retained.
+ * <p>
+ * Implements the {@link BaseHFileCleanerDelegate} for integrating with the HBase cleaner framework
+ * and the {@link Abortable} interface to handle error scenarios gracefully.
+ * </p>
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
+public class ContinuousBackupStagedHFileCleaner extends BaseHFileCleanerDelegate
+  implements Abortable {
+  private static final Logger LOG =
+    LoggerFactory.getLogger(ContinuousBackupStagedHFileCleaner.class);
+
+  private boolean stopped = false;
+  private boolean aborted;
+  private Configuration conf;
+  private Connection connection;
+
+  @Override
+  public void setConf(Configuration config) {
+    this.conf = config;
+    this.connection = null;
+    try {
+      // Establishing HBase connection
+      this.connection = ConnectionFactory.createConnection(conf);
+      LOG.info("HBase connection established successfully.");
+    } catch (IOException ioe) {
+      LOG.error("Couldn't establish connection to HBase", ioe);
+    }
+  }
+
+  @Override
+  public Iterable<FileStatus> getDeletableFiles(Iterable<FileStatus> files) {
+    if (conf == null) {
+      LOG.warn("Configuration is not set. Returning original list of files.");
+      return files;
+    }
+
+    if (connection == null) {
+      try {
+        connection = ConnectionFactory.createConnection(conf);
+        LOG.info("HBase connection re-established in getDeletableFiles.");
+      } catch (IOException e) {
+        LOG.error("Failed to re-establish HBase connection. Returning no deletable files.", e);
+        return Collections.emptyList();
+      }
+    }
+
+    try {
+      // Fetch staged files from HBase
+      Set<String> stagedFiles = StagedBulkloadFileRegistry.listAllBulkloadFiles(connection);
+      LOG.debug("Fetched {} staged files from HBase.", stagedFiles.size());
+
+      // Extract file names from staged files
+      Set<String> stagedFileNames =
+        stagedFiles.stream().map(path -> new Path(path).getName()).collect(Collectors.toSet());
+
+      // Filter files by checking their file names against staged file names
+      return Iterables.filter(files, file -> !stagedFileNames.contains(file.getPath().getName()));
+    } catch (IOException e) {
+      LOG.error("Failed to fetch staged bulkload files from HBase. Returning no deletable files.",
+        e);
+      return Collections.emptyList();
+    }
+  }
+
+  @Override
+  public boolean isFileDeletable(FileStatus fStat) {
+    // The actual deletion decision is made in getDeletableFiles, so returning true
```

Comment:
Why?
Check if the file available for deletion here. Connection to HBase can be kept open.

Reply:
A single `GET` op to HBase checking if filename is present in the table should be quick.

Reply:
This approach would be slightly faster, right? In `getDeletableFiles`, we could run the query once and process all the results together, instead of executing multiple queries.

Reply:
I'm not sure about that. Currently you run a scanner while with my approach, you'll run multiple GETs. Generally speaking PUTs and GETs are the most performant operations in key-value stores, therefore they're preferred over scanners. But. The table is small, cached, etc. 

Personally I don't like scanning the same table over and over again. It's like a Hashtable.